### PR TITLE
tests: enable the code coverage for qemu_x86_64

### PR DESCRIPTION
**Pull request by enjiamai**
Enable the code coverage report for qemu_x86_64. See issue #17991.
In order to make this work, do some order change in linking file (include/arch/x86/intel64/linker.ld ).

Signed-off-by: Enjia Mai <enjiax.mai@intel.com>

### CODE REVIEW

File: boards/x86/qemu_x86/qemu_x86_64_coverage.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+identifier: qemu_x86_64_coverage
```
Comment (nashif) [inline boards/x86/qemu_x86/qemu_x86_64_coverage.yaml:1]: why i sthis  platform definition needed?

Comment (enjiamai) [reply boards/x86/qemu_x86/qemu_x86_64_coverage.yaml:1]: Because there was already qemu_x86_64 identifier exist, that can let us decide whether running with coverage report or not. I try to mimic similar configurations from qemu_x86 and qemu_x86_coverage.  
we can run command as follow:
west build tests/kernel/ - b qemu_x86_64_coverage    or    sanitycheck -T tests/kernel/ -b qemu_x86_64  --coverage
thanks!

Comment (nashif) [reply boards/x86/qemu_x86/qemu_x86_64_coverage.yaml:1]: qemu_x86_coverage is there for different reasons, we do not need to add a special coverage platform for each platform we enable coverage

Comment (enjiamai) [reply boards/x86/qemu_x86/qemu_x86_64_coverage.yaml:1]: Ok, understand. I'll see other example and correct this, thanks!

**Review by LeiW000 — APPROVED** [review]

File: boards/x86/qemu_x86/qemu_x86_64_coverage.yaml (hunk 1 above)
Comment (enjiamai) [reply boards/x86/qemu_x86/qemu_x86_64_coverage.yaml]: I already update this, could you please help to review it again, @nashif  thanks!

File: include/arch/x86/intel64/linker.ld [hunk 2]
```diff
@@ -107,15 +107,16 @@ SECTIONS
...
 #ifdef CONFIG_X86_MMU
 	. = ALIGN(8);
 	_mmu_region_list_start = .;
 	KEEP(*("._mmu_region.static.*"))
 	_mmu_region_list_end = .;
 #endif /* CONFIG_X86_MMU */
 
+/* Move this section after mmu region in order to fix error for running GCOV.*/
+#include <snippets-rodata.ld>
+
```
Comment (jenmwms) [inline include/arch/x86/intel64/linker.ld:119]: Would this be more appropriate as a separate commit? It resolves an error for running GCOV for any x86 with MMU config, not specific to enabling code coverage for qemu_x86_64. Or am I mistaken?

File: kernel/init.c [hunk 3]
```diff
@@ -121,7 +121,7 @@ void z_bss_zero(void)
 #endif	/* CONFIG_CODE_DATA_RELOCATION */
 #ifdef CONFIG_COVERAGE_GCOV
 	(void)memset(&__gcov_bss_start, 0,
-		 ((uint32_t) &__gcov_bss_end - (uint32_t) &__gcov_bss_start));
+		 (uintptr_t) &__gcov_bss_end - (uintptr_t) &__gcov_bss_start);
```
Comment (jenmwms) [inline kernel/init.c:124]: Was the protective outer parenthesis violating coding guidelines? jw, I like to use them but may be wrong to.

File: tests/kernel/mem_protect/mem_map/src/main.c [hunk 4]
```diff
@@ -12,6 +12,13 @@
 #define SKIP_EXECUTE_TESTS
 #endif
 
+/* The execution case in qemu_x86_64 is always failed when running with
+ * coverage option, so skip it temporarily.
+ */
+#if defined(CONFIG_X86_64) && defined(CONFIG_COVERAGE)
+#define SKIP_EXECUTE_TESTS
+#endif
+
```
Comment (jenmwms) [inline tests/kernel/mem_protect/mem_map/src/main.c:21]: If we skip executing the tests, do we still generate coverage reports? (Is this sufficient to close the bug #17991?)

File: include/arch/x86/intel64/linker.ld (hunk 2 above)
Comment (andrewboie) [reply include/arch/x86/intel64/linker.ld:119]: yes the linker.ld changes need to be their own commit

File: tests/kernel/mem_protect/mem_map/src/main.c [hunk 5]
```diff
@@ -12,6 +12,13 @@
 #define SKIP_EXECUTE_TESTS
 #endif
 
+/* The execution case in qemu_x86_64 is always failed when running with
+ * coverage option, so skip it temporarily.
```
Comment (andrewboie) [inline tests/kernel/mem_protect/mem_map/src/main.c:16]: x86_64 has the NX bit, there is no technical reason to disable the execution tests.
Where is the bug on this? Have you made any attempt to root-cause this?

Either way you can't just put "skip it temporarily" with no follow-up, there must be a bug filed, else this will never be fixed.

**Review by tyrantlizard77 — DISMISSED** [review]
Please separate out these changes into logical commits.

File: include/arch/x86/intel64/linker.ld (hunk 2 above)
Comment (enjiamai) [reply include/arch/x86/intel64/linker.ld:119]: @jenmwms @andrewboie  This change is also a part of enable enabling code coverage for qemu_x86_64. And I will split them into 2 commit, thank for your comment~

File: kernel/init.c (hunk 3 above)
Comment (enjiamai) [reply kernel/init.c:124]: @jenmwms You are totally right, I omitted this, and it might be necessary for some safety coding guideline, thanks!

File: tests/kernel/mem_protect/mem_map/src/main.c (hunk 5 above)
Comment (enjiamai) [reply tests/kernel/mem_protect/mem_map/src/main.c:16]: @andrewboie Sorry, I had tried to find the root cause for this couple days but still not know exactly why. It failed due to the fatal error reason is "". The address of <__gcov0.transplanted_function> inside transplanted_function( ) is not correct. It should be like 0xbfffb000,  not should be like 0xc0022dbb. 

=== pass log in qemu_x86 with coverage or qemu_x86_64
START - test_z_mem_map_exec
E: Page fault at address 0xbfffb000 (error code 0x11)
E: Access violation: supervisor thread not allowed to execute

001006be <transplanted_function>:
  1006be:       55                      push   %ebp
  1006bf:       89 e5                   mov    %esp,%ebp 
  1006c1:       a1 18 85 12 00          mov    0x128518,%eax   <=   correct address 0xbfffb000
  1006c6:       8b 15 1c 85 12 00       mov    0x12851c,%edx
  1006cc:       83 c0 01                add    $0x1,%eax


===  fail log in qemu_x86_64 with coverage
START - test_z_mem_map_exec
E: Page fault at address 0xc0022dbb (error code 0x0)  
E: Linear address not present in page tables
E: Access violation: supervisor thread not allowed to read

0000000000100265 <transplanted_function>:
  100265:       55                      push   %rbp
  100266:       48 89 e5                mov    %rsp,%rbp
  100269:       48 89 7d f8             mov    %rdi,-0x8(%rbp)
  10026d:       48 8b 05 ac ad 02 00    mov    0x2adac(%rip),%rax        # 12b020 <__gcov0.transplanted_function> <=   wrong address 0xc0022dbb.

I guess this seems like some thing related to QEMU. And I will keep digging on this.  Thanks for your review and comment!

Comment (andrewboie) [reply tests/kernel/mem_protect/mem_map/src/main.c:16]: Please at least ensure a bug is filed. Thanks!

Comment (enjiamai) [reply tests/kernel/mem_protect/mem_map/src/main.c:16]: @andrewboie Ok, got it. Now I understand to issue a bug first when when met this sort of situation. Thanks for your reminder!

File: tests/kernel/mem_protect/mem_map/src/main.c (hunk 4 above)
Comment (enjiamai) [reply tests/kernel/mem_protect/mem_map/src/main.c:22]: Yes, it just skip this one test case and the coverage report can be generate correctly.

**Comment by enjiamai** [comment]
@andrewboie @jenmwms I have already fix all the comments and suggestion from you, thanks for your great opinion.
And I try to analyze why the memory map execution test case would fail here, then also file a issue for the reason of what causes this case failed. Please see https://github.com/zephyrproject-rtos/zephyr/issues/30434 .
Could your please help to review it again? thanks!

**Review by nashif — APPROVED** [review]

**Review by jenmwms — APPROVED** [review]

**Comment by LeiW000** [comment]
@andrewboie , do you still have any concern on it? If not, I think the PR can be merged now.

**Comment by enjiamai** [comment]
Excuse me, @andrewboie could you please help to review again? thanks!

File: subsys/testsuite/coverage/coverage_ram.ld [hunk 6]
```diff
@@ -31,7 +31,21 @@ __gcov_bss_num_words = ((__gcov_bss_end - __gcov_bss_start) >> 2);
 __gcov_bss_size = __gcov_bss_end - __gcov_bss_start;
 #endif
 
-#ifdef CONFIG_X86
+#ifdef CONFIG_X86_64
```
Comment (finikorg) [inline subsys/testsuite/coverage/coverage_ram.ld:34]: I assume we cannot combine those 2 sections, right?

**Review by finikorg — APPROVED** [review]

File: subsys/testsuite/coverage/coverage_ram.ld (hunk 6 above)
Comment (enjiamai) [reply subsys/testsuite/coverage/coverage_ram.ld:34]: @finikorg  thanks for you review. Yes, I think we better split them up.
